src/NHibernate/Cache/CacheBase.cs

```diff
@@ -0,0 +1,386 @@
+using System.Collections;
+using System.Threading;
+using System.Threading.Tasks;
+using NHibernate.Cache.Entry;
+using NHibernate.Type;
+
+namespace NHibernate.Cache
+{
+	/// <summary>
+	/// Base class for letting implementors define a caching algorithm.
+	/// </summary>
+	/// <remarks>
+	/// <threadsafety instance="true" />
+	/// <para>
+	/// All implementations <em>must</em> be threadsafe.
+	/// </para>
+	/// <para>
+	/// The key is the identifier of the object that is being cached. The key is in most cases
+	/// a <see cref="CacheKey" />.
+	/// </para>
+	/// <para>
+	/// The value can be a <see cref="CacheEntry"/>, a <see cref="CollectionCacheEntry"/>,
+	/// a <see cref="AnyType.ObjectTypeCacheEntry"/>, an <see cref="IList"/> or
+	/// <see cref="IDictionary"/> implementation, all containing simple values or array of
+	/// simple values. It can also be directly a simple value or an array of simple values.
+	/// And it can be a <see cref="CachedItem"/> containing any of the previous types, or
+	/// a <see cref="CacheLock"/>.
+	/// </para>
+	/// <para>
+	/// All those types are binary serializable.
+	/// </para>
+	/// <para>
+	/// This base class provides minimal async method implementations delegating their work to their
+	/// synchronous counterparts. Override them for supplying actual async operations.
+	/// </para>
+	/// <para>
+	/// Similarly, this base class provides minimal multiple get/put/lock/unlock implementations
+	/// delegating their work to their single operation counterparts. Override them if your cache
+	/// implementation supports multiple operations.
+	/// </para>
+	/// </remarks>
+	public abstract partial class CacheBase :
+		// 6.0 TODO: remove ICache
+#pragma warning disable 618
+		ICache
+#pragma warning restore 618
+	{
+		/// <summary>
+		/// A reasonable "lock timeout".
+		/// </summary>
+		public abstract int Timeout { get; }
+
+		/// <summary>
+		/// The name of the cache region.
+		/// </summary>
+		public abstract string RegionName { get; }
+
+		/// <summary>
+		/// Should batched get operations be preferred other single get calls?
+		/// </summary>
+		/// <remarks>
+		/// <para>
+		/// This property should yield <see langword="false" /> if <see cref="GetMany" /> delegates
+		/// its implementation to <see cref="Get"/>.
+		/// </para>
+		/// <para>
+		/// When <see langword="true" />, NHibernate will attempt to get other non initialized proxies or
+		/// collections from the cache instead of only getting the proxy or collection which initialization
+		/// is asked for. If this cache implementation does not benefit from batching together get operations,
+		/// this may result in a performance loss.
+		/// </para>
+		/// <para>
+		/// When <see langword="false" />, NHibernate will still call <see cref="GetMany" /> when it has many
+		/// gets to perform. Its <see cref="CacheBase" /> default implementation is adequate for this case.
+		/// </para>
+		/// </remarks>
+		public abstract bool PreferMultipleGet { get; }
```

Comment:
When the cache does not benefit from multiple gets, NHibernate should not attempt on a single get to pro-actively get up to `batch-size` entries from the cache. Otherwise it would incurs an overhead without any benefit. This flag allows to tell NHibernate if it should pro-actively batch single gets from cache. When doing so, NHibernate may load unneeded data from the cache (the same way `batch-size` can cause NHibernate to load unneeded data from the database), so it is important to avoid doing this for caches not benefiting of multiple gets.

A cache implementation supporting multiple gets could also choose to let users configure this, if the user does not wish the `batch-size` mapping attribute to be applied to fetches from cache.

We do not need a similar flag for put or lock cases, since in these cases NHibernate do no guesses about which items are to be handled. It actually have to cache/lock them all.

Reply:
I think this needs to be `virtual` and yield `false` by default.

Reply:
I have chosen to have it abstract for avoiding it to be overlooked by implementors. Why not defaulting to `false`, but then I will surely add some notes in `GetMany` xml documentation for hinting at caring about this property.